### PR TITLE
Css 10675/authz middleware

### PR DESCRIPTION
**Pull request by SimoneDutto**
## Description

This pr implements a authorization middleware for user/model permissions.

Fixes _JIRA/GitHub issue number_

## Engineering checklist
*Check only items that apply*

- [ ] Documentation updated
- [ ] Covered by unit tests
- [ ] Covered by integration tests

### CODE REVIEW

File: internal/middleware/authz.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+		}
+		modelTag := names.NewModelTag(modelUUID)
+		switch accessNeeded {
+		case ofganames.ReaderRelation:
+			ok, err := user.IsModelReader(ctx, modelTag)
+			if !ok || err != nil {
+				w.WriteHeader(http.StatusForbidden)
+				_, _ = w.Write([]byte("no access to the resource"))
+				return
+			}
```
Comment (kian99) [inline internal/middleware/authz.go:40]: You could wrap this up in a closure to reduce the duplication but like this good too.

File: internal/middleware/authz_test.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := qt.New(t)
+			req := httptest.NewRequest(http.MethodGet, "/", nil)
+			w := httptest.NewRecorder()
+			handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(http.StatusOK)
+			})
+			var h http.Handler
+			if tt.permissionRequired == "reader" {
```
Comment (kian99) [inline internal/middleware/authz_test.go]: Why not just make `tt.permissionRequired` of type ofganames.Relation?

**Review by kian99 — DISMISSED** [review]

File: internal/middleware/authz_test.go (hunk 2 above)
Comment (SimoneDutto) [reply internal/middleware/authz_test.go]: lol, you are right

**Review by kian99 — DISMISSED** [review]
Size: S

File: internal/middleware/authz.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	"github.com/canonical/jimm/v3/internal/jujuapi"
+	ofganames "github.com/canonical/jimm/v3/internal/openfga/names"
+)
+
+// AuthorizeUserForModelAccess extract the user from the context, and checks for permission on the model uuid extracted from the path.
+func AuthorizeUserForModelAccess(next http.Handler, jimm jujuapi.JIMM, accessNeeded cofga.Relation) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		user, err := IdentityFromContext(ctx)
+		if err != nil {
```
Comment (ale8k) [inline internal/middleware/authz.go:21]: Some space between blocks would be nice, it's frustrating to read when all bulked together... i.e.,:

```go
		modelUUID := chi.URLParam(r, "uuid")
		if modelUUID == "" {
			w.WriteHeader(http.StatusUnauthorized)
			_, _ = w.Write([]byte("cannot find uuid in URL path"))
			return
		}

		modelTag := names.NewModelTag(modelUUID)

		switch accessNeeded {
		case ofganames.ReaderRelation:
```

**Review by ale8k — APPROVED** [review]
lgtm

**Comment by ale8k** [comment]
size s

File: internal/middleware/authz.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+			if !ok || err != nil {
+				w.WriteHeader(http.StatusForbidden)
+				_, _ = w.Write([]byte("no access to the resource"))
+				return
+			}
+		case ofganames.WriterRelation:
+			ok, err := user.IsModelWriter(ctx, modelTag)
+			if !ok || err != nil {
+				w.WriteHeader(http.StatusForbidden)
+				_, _ = w.Write([]byte("no access to the resource"))
```
Comment (kian99) [inline internal/middleware/authz.go:45]: nit: technically they might have access to the resource but without the right level.
Maybe "insufficient access to the resource".
And if you really want to take it far, if they have no access at all, for more security we should return a 404 and if they don't have enough access we should return a 403. But I'm happy with the way it is currently.

**Review by kian99 — APPROVED** [review]

**Pull request merged**